### ISSUE TITLE
Expose more leaderboard query options

### ISSUE THREAD

**Issue by EIREXE** [heroiclabs/nakama#427]
Currently leaderboard querying seems overly simplistic, and it isn't possible to query the leaderboard list (as far as I can see).

Here's my use case:

I develop a rhythm game called Project Heartbeat, in Project Heartbeat each community-made song chart is assigned a guid on creation that I would like for the server to use to identify the song. The problem is that songs can have modifiers added to them, thus multiple leaderboards for the same song would be created or it would be done in other ways.

The way Osu does this is they have what I think is a leaderboard for each song and they query depending on settings given by the users, some modifiers increase score and others decrease it.

This seems awfully inefficient to me but it works for them.

So to sum up, it would be nice to have a way to filter by leaderboard metadata or to query the list of leaderboards or even better to have leaderboard groups.

I could work on a PR, if you consider this is somethin within the scope of nakama, or build my own leaderboard system as runtime code.

Cheers

**Comment by novabyte** [comment]
@EIREXE Thanks for the detailed feature request. It's good to see the use case described from the perspective of game feature. Thanks.

> Currently leaderboard querying seems overly simplistic, and it isn't possible to query the leaderboard list (as far as I can see).

This is definitely something which would be useful to add. We originally designed the leaderboard APIs to assume that the identifier is known ahead of time from a separate source of truth (config file, guild ID, etc, etc). 

> So to sum up, it would be nice to have a way to filter by leaderboard metadata or to query the list of leaderboards or even better to have leaderboard groups.

We implemented this approach in the past and it gets complicated fast. Also performance becomes more variable because some score updates will be persisted to multiple leaderboards underneath the leaderboard group. 

A better approach is to keep the leaderboards themselves loosely coupled and use some other storage object to maintain each of the "groupings" based on the specifics of the game design.

-- 

A pull request to allow leaderboards to be listed would be great. I think its important to know what the filter criteria you'd like to support on that table looks like ahead of the implementation kickoff though. What do you have in mind?

**Comment by EIREXE** [comment]
> @EIREXE Thanks for the detailed feature request. It's good to see the use case described from the perspective of game feature. Thanks.
> 
> > Currently leaderboard querying seems overly simplistic, and it isn't possible to query the leaderboard list (as far as I can see).
> 
> This is definitely something which would be useful to add. We originally designed the leaderboard APIs to assume that the identifier is known ahead of time from a separate source of truth (config file, guild ID, etc, etc).
> 
> > So to sum up, it would be nice to have a way to filter by leaderboard metadata or to query the list of leaderboards or even better to have leaderboard groups.
> 
> We implemented this approach in the past and it gets complicated fast. Also performance becomes more variable because some score updates will be persisted to multiple leaderboards underneath the leaderboard group.
> 
> A better approach is to keep the leaderboards themselves loosely coupled and use some other storage object to maintain each of the "groupings" based on the specifics of the game design.
> 
> --
> 
> A pull request to allow leaderboards to be listed would be great. I think its important to know what the filter criteria you'd like to support on that table looks like ahead of the implementation kickoff though. What do you have in mind?

How would you suggest this be done architecture-wise? I was considering just being able to filter by leaderboard meta, but this might not be very robust/fast. You clearly have more experience by me so I am not sure if this is something that can be made generic so anyone can use it or if it's a specific customization I should do for my case.

I will be looking into being able to query for leaderboard search.

**Comment by novabyte** [comment]
@EIREXE Simplest support to add is wildcard prefix searches on the leaderboard ID. For example `WHERE id ILIKE 'song1%'`. For more complex criteria I think we should discuss the types of filters you need for the game in more detail here.

The prefix search would perform well because the leaderboard ID is the primary key of that table so we already have a unique index on it.

**Comment by EIREXE** [comment]
> @EIREXE Simplest support to add is wildcard prefix searches on the leaderboard ID. For example `WHERE id ILIKE 'song1%'`. For more complex criteria I think we should discuss the types of filters you need for the game in more detail here.
> 
> The prefix search would perform well because the leaderboard ID is the primary key of that table so we already have a unique index on it.

In my case, each leaderboard entry would also contain information on what song modifiers were be used for playing back the song, for example if someone has chosen the song speed up modifier and set it to 125%. In my case I would like to be able to filter by song modifier.
